### ISSUE TITLE
Editor is very slow when one component instance or gui template is used in many places

### ISSUE THREAD

**Issue by notbadgun** [defold/defold#5447]
### Expected behaviour
I’m developing a physics game, and I make levels in Defold IDE. Each level is a collection with many pre-made game objects (obstacles, enemies, bonuses, walls). It’s a very convenient way to make levels because I can preview my levels right in the IDE and I don’t have to spend my time to make a special level editor.

How it looks:
![0bb150f22a0953e1ab65390010df5e031311d097](https://user-images.githubusercontent.com/13905665/103519438-9528df00-4e86-11eb-8857-4245f1d8a3d2.png)

My game will have 250+ levels. I made 60 levels, and it’s a pain to make more.

### Actual behaviour
IDE is very slow when there are many collections

### Steps to reproduce
For example, you can try to change something or to edit the main.script, i.e. even entering new text is lagging on my PC for 0.5-1 sec. PC has 8 core Ryzen 2700X.

**Demo project:**
[lot_of_collections1.zip](https://github.com/defold/defold/files/5763832/lot_of_collections1.zip)

**+1**
https://twitter.com/deulamco/status/1345740345900900352

<hr/>

<table>
<tr><td>Build time</td><td>2020-12-18T20:28:49.967999</td></tr>
<tr><td>Defold channel</td><td>editor-alpha</td></tr>
<tr><td>Defold editor sha</td><td>15990df1280c63c1744849d6456237c6411f593e</td></tr>
<tr><td>Defold engine sha</td><td>f7778a8f59ef2a8dda5d445f471368e8bd1cb1ac</td></tr>
<tr><td>Defold version</td><td>1.2.177</td></tr>
<tr><td>GPU</td><td>GeForce GTX 1080 Ti/PCIe/SSE2</td></tr>
<tr><td>GPU Driver</td><td>4.6.0 NVIDIA 460.89</td></tr>
<tr><td>Java version</td><td>11.0.1+13</td></tr>
<tr><td>OS arch</td><td>amd64</td></tr>
<tr><td>OS name</td><td>Windows 10</td></tr>
<tr><td>OS version</td><td>10.0</td></tr>
<table>

**Comment by subsoap** [comment]
I think this is part of a problem of non-localized state.

**Comment by britzl** [comment]
From @matgis 

There is a concept called successors in the graph, which is the recursive set of outputs that are upstream of an output. For example, the build-targets output of a Script node is a successor of its own lines output, as is every output in all other nodes that are affected by the build-targets output, and so on.
For a script that is referenced a lot, there will be a lot of successors for the lines output. When it slows down, it seems we’re spending a lot of time just evicting all the potentially cached values for the successors from the cache.
I think the way forward is to be smarter about how we evict values from the cache. Instead of always evicting the full set of recursive successors every time an output is invalidated, we could perhaps do it in a recursive manner that could early-out if we know that no upstream value that pulls from a higher-level successor has entered the cache.

**Comment by totebo** [comment]
I've bumped into this issue too. In layman's terms, the slowdown happens when the file edited is used in many other places. In my case, object.script exists in every object in the game, and the game has 100 levels. Editing object.script in Defold is very slow, so I have started editing it in an external editor, then coming back to Defold and wait for the changes to propagate.

**Comment by AGulev** [comment]
Similar situation with GUI and GUI templates. If there are many GUIs in a big project where the same template used it takes too long to make any change in this template, and editing maуbe a huge problem.

I think it's the same issue, what do you think @britzl ?

**Comment by britzl** [comment]
> I think it's the same issue, what do you think @britzl ?

Yes, I agree!

**Comment by vlaaad** [comment]
Ideas to explore:
- store endpoint predecessors in addition to successors, make incrementing invalidation counters affect only actually changed outputs, then make invalidation counter check go through all predecessors instead of a single lookup. This will increase memory consumption and move CPU load from the "apply transaction" place to the "fill cache" place that can happen asynchronously. This change might improve the performance by up to 50% (will remove the invalidate-counters step).
- only return `:cached` outputs in `basis-dependencies`, I did some experiments with this and the speed up was maybe around 30%, I didn't commit this because this change breaks the contract of `basis-dependencies`, though it might be used only for cache now;
- parallelize calculation of `basis-dependencies`, I didn't evaluate how much this will improve the performance, if it even will improve things.
- maybe find better data structures to store successors? e.g. int map... I think right now the main problem with calculating dependencies is that we navigate big data structures many times.

**Comment by vlaaad** [comment]
Some notes:
1. The predecessors idea is problematic because the cost of inserting cache entry for a collection that references a lot of resources will be similarly very high, so we just move the problem to another place. One thing that might work is some mix of successors and predecessors, i.e. use both, and on transact make encaching process look at all predecessors plus enqueue an async successors update that will make encaching process look at only the individual cache entry counters again. Not sure how hard would it be to implement this synchronization.
2. We might get some improvements in performance if we reduce the number of connections. For example, we could express dep build targets as some "implied" connections, e.g. just resources instead of resource nodes, and then the build process will follow these implied connections so the behavior is the same.

**Comment by vlaaad** [comment]
Exploring successors optimization... There are 3 types of connections:
- internal, defined as dependencies inside a single node
- override, defined as a dependency of a node's label on the original node's label of the same name
- external, a connection between 2 different nodes in the graph

There seem to be 2 areas to explore:
1. optimizing internal connections of highly-referenced nodes. This will affect both internal and override successors.
2. optimizing the topology of nodes by expressing dependencies outside of the graph. If we will be able to find topology optimizations, this will affect everything, since removing external links will also remove internal/override connections of externally-referenced nodes. 

I investigated where to look for internal connection optimizations, here is a breakdown of connections by node type and successor type for highly-referenced GUI node:
```
total 13150 nodes, 198137 successors
successor contribution:
override (22.86%, 45299 total)
    8.28% GuiSceneNode (16409 total)
    7.82% BoxNode (15493 total)
    3.55% TemplateNode (7032 total)
    2.81% NodeTree (5576 total)
    0.29% TextNode (575 total)
    0.09% SpineNode (187 total)
    0.01% ParticleFXNode (26 total)
    0.00% PieNode (1 total)
internal (43.20%, 85592 total)
   22.30% GuiSceneNode (44179 total)
   11.33% BoxNode (22456 total)
    5.18% TemplateNode (10270 total)
    3.36% NodeTree (6649 total)
    0.57% TextNode (1130 total)
    0.18% SpineNode (359 total)
    0.12% EmbeddedGOInstanceNode (234 total)
    0.05% ReferencedComponent (104 total)
    0.03% EmbeddedComponent (54 total)
    0.03% ParticleFXNode (51 total)
    0.03% PieNode (50 total)
    0.01% SceneView (28 total)
    0.01% CurveView (14 total)
    0.00% SceneVisibilityNode (8 total)
    0.00% AppView (3 total)
    0.00% Project (2 total)
    0.00% Grid (1 total)
external (33.94%, 67246 total)
    8.81% NodeTree->GuiSceneNode (17448 total)
    6.06% BoxNode->BoxNode (12002 total)
    4.89% GuiSceneNode->TemplateNode (9696 total)
    3.97% TemplateNode->BoxNode (7870 total)
    2.99% BoxNode->NodeTree (5931 total)
    2.63% TemplateNode->GuiSceneNode (5220 total)
    1.52% BoxNode->TemplateNode (3012 total)
    0.55% GuiSceneNode->NodeTree (1083 total)
    0.50% NodeTree->BoxNode (996 total)
    0.50% BoxNode->TextNode (984 total)
    0.24% LayoutNode->GuiSceneNode (480 total)
    0.18% BoxNode->SpineNode (353 total)
    0.16% GameObjectNode->EmbeddedGOInstanceNode (321 total)
    0.14% GuiSceneNode->Project (270 total)
    0.12% TemplateNode->NodeTree (240 total)
    0.09% EmbeddedGOInstanceNode->CollectionNode (177 total)
    0.09% GuiSceneNode->ReferencedComponent (175 total)
    0.07% NodeTree->TextNode (132 total)
    0.06% ReferencedComponent->GameObjectNode (123 total)
    0.06% NodeTree->TemplateNode (114 total)
    0.04% FactoryNode->EmbeddedComponent (86 total)
    0.04% NodeTree->LayoutNode (80 total)
    0.03% EmbeddedComponent->GameObjectNode (54 total)
    0.02% BoxNode->ParticleFXNode (49 total)
    0.02% CollectionNode->FactoryNode (43 total)
    0.02% PieNode->PieNode (40 total)
    0.01% TextNode->BoxNode (28 total)
    0.01% PieNode->BoxNode (22 total)
    0.01% CollectionProxyNode->EmbeddedComponent (22 total)
    0.01% TemplateNode->PieNode (20 total)
    0.01% GuiSceneNode->LayoutsNode (19 total)
    0.01% LayoutsNode->GuiSceneNode (19 total)
    0.01% BoxNode->PieNode (17 total)
    0.01% TemplateNode->TextNode (14 total)
    0.01% GuiSceneNode->SceneView (14 total)
    0.01% CameraController->SceneView (13 total)
    0.01% TextNode->NodeTree (12 total)
    0.01% CollectionNode->CollectionProxyNode (11 total)
    0.01% AppView->CurveView (10 total)
    0.00% NodeTree->SpineNode (6 total)
    0.00% TextNode->TemplateNode (6 total)
    0.00% AppView->SceneVisibilityNode (6 total)
    0.00% ToolController->SceneView (5 total)
    0.00% SpineNode->TextNode (4 total)
    0.00% SceneView->AppView (2 total)
    0.00% CameraController->Grid (2 total)
    0.00% Rulers->SceneView (2 total)
    0.00% Grid->SceneView (2 total)
    0.00% NodeTree->ParticleFXNode (2 total)
    0.00% Project->AppView (1 total)
    0.00% CollectionNode->GameProjectNode (1 total)
    0.00% CameraController->Rulers (1 total)
    0.00% SceneView->ToolController (1 total)
    0.00% BoxNode->Project (1 total)
    0.00% SceneView->CameraController (1 total)
    0.00% AppView->PropertiesView (1 total)
    0.00% GuiSceneNode->AppView (1 total)
    0.00% CameraController->ToolController (1 total)
```
Here is a similar report for highly-referenced script file:
```
total 16357 nodes, 572876 successors
successor contribution:
override (26.64%, 152586 total)
    8.44% ScriptNode (48324 total)
    7.73% GameObjectNode (44275 total)
    5.62% ReferencedComponent (32200 total)
    4.56% ReferencedGOInstanceNode (26100 total)
    0.19% CollectionNode (1111 total)
    0.10% CollectionInstanceNode (576 total)
internal (44.18%, 253069 total)
   18.98% ScriptNode (108756 total)
   12.65% GameObjectNode (72486 total)
    7.03% ReferencedComponent (40270 total)
    4.92% ReferencedGOInstanceNode (28175 total)
    0.47% CollectionNode (2698 total)
    0.11% CollectionInstanceNode (606 total)
    0.01% CodeEditorView (54 total)
    0.00% CurveView (14 total)
    0.00% EmbeddedGOInstanceNode (6 total)
    0.00% AppView (2 total)
    0.00% Project (2 total)
external (29.19%, 167221 total)
    7.73% ReferencedComponent->GameObjectNode (44297 total)
    7.73% ReferencedGOInstanceNode->CollectionNode (44275 total)
    7.03% GameObjectNode->ReferencedGOInstanceNode (40250 total)
    6.33% ScriptNode->ReferencedComponent (36243 total)
    0.19% CollectionInstanceNode->CollectionNode (1111 total)
    0.16% CollectionNode->CollectionInstanceNode (909 total)
    0.01% CollectionNode->Project (82 total)
    0.00% ScriptNode->CodeEditorView (15 total)
    0.00% EmbeddedGOInstanceNode->CollectionNode (11 total)
    0.00% AppView->CurveView (10 total)
    0.00% GameObjectNode->EmbeddedGOInstanceNode (8 total)
    0.00% ScriptNode->Project (3 total)
    0.00% GameObjectNode->Project (2 total)
    0.00% CodeEditorView->AppView (2 total)
    0.00% Project->AppView (1 total)
    0.00% CollectionNode->GameProjectNode (1 total)
    0.00% AppView->PropertiesView (1 total)
```
I'll try experimenting with decreasing the number of internal connections next.

**Comment by vlaaad** [comment]
I also explored the topology of the graph view, but visualizations, albeit showing some pattern, didn't really help: https://twitter.com/v1aaad/status/1616496434441818118

Another idea about topology that I found is trying to break all connections of a particular type (e.g. from BoxNode's `:node-msgs` to NodeTree's `:node-msgs`) and see how much it will affect the number of successors. The idea is that removing such links might have a big effect on the successor tree, because removal of this link might add up if a lot of transitive outputs are affected. Here is an output for highly-referenced GUI node:
```
Baseline: 198137 successors
Improvements by pair link class:
  61.66% (of which 0.95% direct) BoxNode:node-msgs -> NodeTree:node-msgs
  49.53% (of which 1.04% direct) GuiSceneNode:pb-msg -> TemplateNode:node-msgs
  46.80% (of which 1.06% direct) TemplateNode:node-msgs -> BoxNode:node-msgs
  38.21% (of which 1.44% direct) BoxNode:node-msgs -> BoxNode:node-msgs
  30.05% (of which 0.03% direct) NodeTree:node-msgs -> LayoutNode:pb-msg
  29.60% (of which 1.85% direct) GuiSceneNode:current-layout -> NodeTree:current-layout
  29.03% (of which 1.73% direct) NodeTree:current-layout -> BoxNode:current-layout
  27.34% (of which 7.74% direct) BoxNode:current-layout -> BoxNode:current-layout
  26.01% (of which 1.95% direct) BoxNode:current-layout -> TemplateNode:current-layout
  24.28% (of which 2.17% direct) TemplateNode:current-layout -> GuiSceneNode:current-layout
  17.31% (of which 3.39% direct) BoxNode:node-rt-msgs -> NodeTree:node-rt-msgs
   7.93% (of which 6.51% direct) GuiSceneNode:rt-pb-msg -> TemplateNode:node-rt-msgs
   7.14% (of which 6.94% direct) TemplateNode:node-rt-msgs -> BoxNode:node-rt-msgs
   6.36% (of which 9.19% direct) NodeTree:node-msgs -> GuiSceneNode:node-msgs
   5.38% (of which 9.58% direct) GuiSceneNode:build-targets -> TemplateNode:template-build-targets
   5.07% (of which 10.82% direct) BoxNode:node-rt-msgs -> BoxNode:node-rt-msgs
   4.18% (of which 0.46% direct) NodeTree:current-layout -> TemplateNode:current-layout
   3.88% (of which 12.76% direct) TemplateNode:template-build-targets -> BoxNode:template-build-targets
   3.08% (of which 0.56% direct) TemplateNode:node-msgs -> NodeTree:node-msgs
   2.85% (of which 20.48% direct) NodeTree:node-rt-msgs -> GuiSceneNode:node-rt-msgs
   2.42% (of which 26.18% direct) BoxNode:scene -> NodeTree:child-scenes
   2.17% (of which 25.36% direct) BoxNode:template-build-targets -> BoxNode:template-build-targets
   2.05% (of which 37.30% direct) GuiSceneNode:scene -> TemplateNode:scene-children
   2.05% (of which 37.30% direct) GuiSceneNode:scene -> TemplateNode:aabb
   1.62% (of which 43.00% direct) BoxNode:scene -> BoxNode:scene-children
   1.60% (of which 46.05% direct) TemplateNode:scene -> BoxNode:scene-children
   1.30% (of which 48.48% direct) NodeTree:scene -> GuiSceneNode:child-scenes
   1.28% (of which 38.77% direct) BoxNode:current-layout -> TextNode:current-layout
   1.25% (of which 26.81% direct) BoxNode:template-build-targets -> NodeTree:template-build-targets
   ...skip results with less than 1% improvement
```
This hints we should investigate moving the following dependencies outside of the graph: 
- `:node-msgs`/`:pb-msg` in GUI nodes (up to 60% improvement)
- `:current-layout` (up to 30% improvement)
- `:node-rt-msgs`/`:rt-pb-msg` (up to 18% improvement)

Here is the output for a highly-referenced script file:
```
Baseline: 572776 successors
Improvements by pair link class:
  18.31% (of which 3.84% direct) ReferencedComponent:ddf-message -> GameObjectNode:ddf-component-properties
   8.11% (of which 8.66% direct) GameObjectNode:ddf-component-properties -> ReferencedGOInstanceNode:ddf-message
   7.47% (of which 9.41% direct) ScriptNode:_properties -> ReferencedComponent:node-outline
   6.33% (of which 11.11% direct) ReferencedComponent:ddf-message -> GameObjectNode:proto-msg
   6.06% (of which 11.59% direct) ReferencedComponent:node-outline -> GameObjectNode:node-outline
   4.92% (of which 14.29% direct) ScriptNode:_properties -> ReferencedComponent:ddf-properties
   4.02% (of which 17.49% direct) ReferencedComponent:build-targets -> GameObjectNode:build-targets
   2.61% (of which 26.88% direct) GameObjectNode:ddf-component-properties -> ReferencedGOInstanceNode:path
   2.61% (of which 26.88% direct) GameObjectNode:node-outline -> ReferencedGOInstanceNode:source-outline
   2.33% (of which 30.10% direct) GameObjectNode:ddf-component-properties -> ReferencedGOInstanceNode:ddf-properties
   2.11% (of which 33.34% direct) ScriptNode:build-targets -> ReferencedComponent:build-resource
   1.91% (of which 36.76% direct) GameObjectNode:build-targets -> ReferencedGOInstanceNode:build-resource
   1.13% (of which 62.45% direct) ReferencedGOInstanceNode:ddf-properties -> CollectionNode:ddf-properties
   ...skip results with less than 1% improvement
```
Here, I don't see there is much we can do, except maybe extracting dependencies of `:ddf-message`/`:ddf-component-properties` outside of the graph, but even that it's at most an 18% improvement.

**Comment by vlaaad** [comment]
As a result of previous investigations, I found a place that got us a 25% performance improvement in the case of highly-referenced GUI nodes.

Here are the updated stats about connection health:
```
Total 6101 nodes, 146366 successors
Successor contribution by type:
  internal (47.53%, 69567 total)
   25.35% GuiSceneNode (37100 total)
   14.72% BoxNode (21545 total)
    3.68% NodeTree (5384 total)
    3.49% TemplateNode (5110 total)
    ...
  external (27.71%, 40551 total)
   11.51% NodeTree -> GuiSceneNode (16854 total)
    4.19% GuiSceneNode -> TemplateNode (6132 total)
    3.72% BoxNode -> BoxNode (5440 total)
    3.63% BoxNode -> NodeTree (5319 total)
    3.35% TemplateNode -> BoxNode (4910 total)
    ...
  override (24.77%, 36248 total)
    9.78% GuiSceneNode (14308 total)
    8.57% BoxNode (12545 total)
    3.22% NodeTree (4710 total)
    3.20% TemplateNode (4680 total)
    ...
```

The difference with [before](https://github.com/defold/defold/issues/5447#issuecomment-1398234021): in absolute terms, every successor type has fewer successors. In relative terms, external is down (33%->27%), internal is up (43%->47%), and override is up (22%->24%)

Stats about external connections contributing to the successors' size:
```
  42.97% (of which 1.85% direct) BoxNode:node-msgs -> NodeTree:node-msgs
  26.54% (of which 2.63% direct) GuiSceneNode:pb-msg -> TemplateNode:node-msgs
  25.25% (of which 2.66% direct) TemplateNode:node-msgs -> BoxNode:node-msgs
  23.44% (of which 3.39% direct) BoxNode:node-rt-msgs -> NodeTree:node-rt-msgs
  18.34% (of which 4.05% direct) BoxNode:node-msgs -> BoxNode:node-msgs
  10.73% (of which 6.51% direct) GuiSceneNode:rt-pb-msg -> TemplateNode:node-rt-msgs
   9.66% (of which 6.94% direct) TemplateNode:node-rt-msgs -> BoxNode:node-rt-msgs
   8.60% (of which 9.19% direct) NodeTree:node-msgs -> GuiSceneNode:node-msgs
   7.29% (of which 9.58% direct) GuiSceneNode:build-targets -> TemplateNode:template-build-targets
   6.87% (of which 10.82% direct) BoxNode:node-rt-msgs -> BoxNode:node-rt-msgs
   5.26% (of which 12.76% direct) TemplateNode:template-build-targets -> BoxNode:template-build-targets
   3.86% (of which 20.48% direct) NodeTree:node-rt-msgs -> GuiSceneNode:node-rt-msgs
   3.07% (of which 25.92% direct) BoxNode:scene -> NodeTree:child-scenes
   2.93% (of which 25.36% direct) BoxNode:template-build-targets -> BoxNode:template-build-targets
   2.27% (of which 34.88% direct) NodeTree:scene -> GuiSceneNode:child-scenes
   1.93% (of which 36.19% direct) GuiSceneNode:scene -> TemplateNode:scene-children
   1.93% (of which 36.19% direct) GuiSceneNode:scene -> TemplateNode:aabb
   1.89% (of which 39.39% direct) BoxNode:scene -> BoxNode:scene-children
   1.69% (of which 26.81% direct) BoxNode:template-build-targets -> NodeTree:template-build-targets
   1.60% (of which 42.00% direct) TemplateNode:scene -> BoxNode:scene-children
   ...
```

As expected, `current-layout` is gone. So now the main culprit is a proliferation of `pb-msg/node-msgs` and `rt-pb-msg/node-rt-msgs`, followed by build targets, now it 7% (up from 5%)

**Comment by vlaaad** [comment]
We did a lot of improvements in this area during the last 10 months. Even though it's still somewhat slow on big projects, it's much better than it was. There are now very few improvements available to us, and most of them will require a lot of exploratory and speculative work, so we decided to not proceed with them for now.
